File: haskell/word-count/WordCount.hs
module WordCount (wordCount) where

import qualified Data.Map.Strict as M (Map, empty, insertWith)
import Data.Char (toLower, isAlphaNum)
import Data.List (foldl')
import Data.List.Split (wordsBy)

type Count = Int
type Word = String
type CountedWords = M.Map Word Count

wordCount :: String -> CountedWords
wordCount str = gatherWordCount M.empty (wordList str)
    where
        wordList = wordsBy (not . isAlphaNum) . map toLower
        gatherWordCount = foldl' (\ counts x -> M.insertWith (+) x 1 counts)

File: javascript/word-count/word-count.js

var Words = function(str) {
	var counts = {};
	var words = str.split(/\s+/);
	words.forEach(function(word) {
		if (counts.hasOwnProperty(word)) {
			counts[word] += 1;
		} else {
			counts[word] = 1;
		}
	});
	return counts;
};

module.exports = Words;